### PR TITLE
Add condition handlers, implement for Degraded prefixing, add StaleConditionController

### PR DESCRIPTION
**Pull request by benjaminapetersen**
Implements part of #260 

Degraded is a pretty significant status in terms of our telemetry, but also infrequently used in code.  Updating first for these reasons.

/assign @jhadvig 
@spadgett

### CODE REVIEW

File: pkg/console/operator/operator.go [hunk 1]
```diff
@@ -240,7 +240,10 @@ func (c *consoleOperator) handleSync(configs configSet) error {
 	err := c.sync_v400(updatedStatus, configs)
 	if err != nil {
 		if !customerrors.IsSyncError(err) || !customerrors.IsCustomLogoError(err) {
-			c.SyncStatus(c.ConditionResourceSyncDegraded(updatedStatus, err.Error()))
+			// TODO: getting the err type like this is not ideal, this mechanism needs rework
```
Comment (benjaminapetersen) [inline pkg/console/operator/operator.go]: This little mechanism has always been a bandaid.  
But, I don't want to do too much here.  Perhaps we will need to backport this fix to 4.1? I don't know if thats necessary or not, so want to keep this PR small enough that its possible should the need arise.

File: pkg/console/operator/operator.go [hunk 2]
```diff
@@ -238,16 +238,21 @@ func (c *consoleOperator) handleSync(configs configSet) error {
 	// set default conditions ok first, sync can toggle if conditions are invalid
 	c.ConditionsDefault(updatedStatus)
 	err := c.sync_v400(updatedStatus, configs)
-	if err != nil {
-		if !customerrors.IsSyncError(err) || !customerrors.IsCustomLogoError(err) {
-			c.SyncStatus(c.ConditionResourceSyncDegraded(updatedStatus, err.Error()))
+
+	c.HandleDegraded(updatedStatus, "SyncError", func() error {
```
Comment (benjaminapetersen) [inline pkg/console/operator/operator.go]: These handler can't be hidden behind `if` statements, else we will be pinned in a permanent condition.  The handler deals with both `true` & `false`.  Thus this block of logic needs to change a bit.

File: pkg/console/operator/status.go [hunk 3]
```diff
@@ -119,15 +173,6 @@ func (c *consoleOperator) SetStatusCondition(operatorConfig *operatorsv1.Console
 	return operatorConfig
 }
 
-func (c *consoleOperator) ConditionResourceSyncSuccess(operatorConfig *operatorsv1.Console) *operatorsv1.Console {
```
Comment (benjaminapetersen) [inline pkg/console/operator/status.go]: All of these helpers will go away in follow-up PRs.

File: pkg/console/operator/sync_v400.go [hunk 4]
```diff
@@ -86,10 +86,10 @@ func (co *consoleOperator) sync_v400(updatedOperatorConfig *operatorv1.Console,
 	if customLogoError != nil {
 		msg := fmt.Sprintf("%q: %v", "customlogoconfigmap", customLogoError)
 		klog.V(4).Infof("incomplete sync: %v", msg)
-		// If the custom logo sync fails for any reason, we are degraded, not progressing.
-		// The sync loop may not settle, we are unable to honor it in current state.
-		co.ConditionDegraded(updatedOperatorConfig, "CustomLogoInvalid", msg)
 	}
+	// If the custom logo sync fails for any reason, we are degraded, not progressing.
```
Comment (benjaminapetersen) [inline pkg/console/operator/sync_v400.go]: hanlder can't be hidden behind an `if`, unlike the logging.

**Comment by benjaminapetersen** [comment]
Expected failure:

```golang
waited 90 seconds to reach settled state...
waited 120 seconds to reach settled state...
--- FAIL: TestOperatorConfigBranding (252.46s)
    brand_builtins_test.go:67: update operator with okd
    brand_builtins_test.go:67: update operator with ocp
    brand_builtins_test.go:67: update operator with online
    brand_builtins_test.go:67: update operator with dedicated
    brand_builtins_test.go:67: update operator with azure
    brand_builtins_test.go:50: operator has not reached settled state in 4m0s attempts due to [Degraded Degraded] - timed out waiting for the condition
```
I need to update the e2e to handle the indirect setting of `Degraded`.

File: pkg/console/operator/status.go [hunk 5]
```diff
@@ -57,6 +58,59 @@ const (
...
+		return
+	}
+	v1helpers.SetOperatorCondition(&operatorConfig.Status.Conditions,
+		operatorsv1.OperatorCondition{
+			Type:   conditionType,
+			Status: operatorsv1.ConditionFalse,
+		})
+}
+
+func IsDegraded(operatorConfig *operatorsv1.Authentication) bool {
```
Comment (jhadvig) [inline pkg/console/operator/status.go]: Can't find where this func is used

File: pkg/console/operator/status.go [hunk 6]
```diff
@@ -176,51 +221,7 @@ func (c *consoleOperator) ConditionResourceSyncNotProgressing(operatorConfig *op
...
-		Type:    operatorsv1.OperatorStatusTypeDegraded,
-		Status:  operatorsv1.ConditionTrue,
-		Message: message,
-		Reason:  reasonSyncError,
-	})
-
-	return operatorConfig
-}
-
+// TODO: potentially eliminate the defaulting mechanism
```
Comment (jhadvig) [inline pkg/console/operator/status.go]: 👍

File: pkg/console/starter/starter.go [hunk 7]
```diff
@@ -176,6 +177,20 @@ func RunOperator(ctx *controllercmd.ControllerContext) error {
 		ctx.EventRecorder,
 	)
 
+	staleConditionsController := staleconditions.NewRemoveStaleConditions(
```
Comment (jhadvig) [inline pkg/console/starter/starter.go]: 👍

File: pkg/console/operator/sync_v400.go [hunk 8]
```diff
@@ -86,10 +86,10 @@ func (co *consoleOperator) sync_v400(updatedOperatorConfig *operatorv1.Console,
 	if customLogoError != nil {
 		msg := fmt.Sprintf("%q: %v", "customlogoconfigmap", customLogoError)
 		klog.V(4).Infof("incomplete sync: %v", msg)
```
Comment (jhadvig) [inline pkg/console/operator/sync_v400.go]: eventually the logging could be part of the `handleCondition `

File: pkg/console/operator/status.go [hunk 9]
```diff
@@ -234,12 +235,8 @@ func (c *consoleOperator) ConditionsDefault(operatorConfig *operatorsv1.Console)
...
 		Message: "As expected",
 	})
-	v1helpers.SetOperatorCondition(&operatorConfig.Status.Conditions, operatorsv1.OperatorCondition{
-		Type:    operatorsv1.OperatorStatusTypeDegraded,
-		Status:  operatorsv1.ConditionFalse,
-		Reason:  reasonAsExpected,
-		Message: "As expected",
-	})
+
+	c.HandleDegraded(operatorConfig, "Default", nil)
```
Comment (jhadvig) [inline pkg/console/operator/status.go]: Looks promising 👍 We will need to revise the `consoleOperator`'s  `Condition*` methods...

File: pkg/console/operator/operator.go (hunk 2 above)
Comment (jhadvig) [reply pkg/console/operator/operator.go]: It kinda dost but this pretty good case to use it :)

File: pkg/console/operator/status.go (hunk 9 above)
Comment (benjaminapetersen) [reply pkg/console/operator/status.go]: Yup. will continue on and hit all of these.  Is a bugzilla now, the condition statuses are not good enough.

File: pkg/console/operator/sync_v400.go (hunk 8 above)
Comment (benjaminapetersen) [reply pkg/console/operator/sync_v400.go]: Thats not a bad idea.  I dislike all the noise everywhere. We should factor the logging noise out.

File: pkg/console/operator/status.go (hunk 5 above)
Comment (benjaminapetersen) [reply pkg/console/operator/status.go]: Neither are the `HandleProgressing()` and `HandleAvailable()`, but I put all these helpers in this first PR so I can build on them.

**Comment by benjaminapetersen** [comment]
Tests ran locally, all passed.  hopefully will be in the clear.

File: pkg/http/routes.go [hunk 10]
```diff
@@ -0,0 +1,24 @@
+package http
```
Comment (jhadvig) [inline pkg/http/routes.go]: not sure why this is here 😄

**Pull request merged**